Re-unify module and interface in ConcreteModuleType

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29826 Fix type sharing on loaded ScriptModules
* **#29825 Re-unify module and interface in ConcreteModuleType**
* #29824 Split ConcreteModuleType into two types

We made `ModuleInfo` a union initially to represent the idea that a
submodule could either be a regular module or a module interface.

This PR represents module interfaces as a ConcreteModuleType with no
info (e.g.  no "sugaredness"), and with the interface type as the
underlying `jitType_`. This has the effect of reducing the special
casing around adding/maintaining module info.

Differential Revision: [D18575011](https://our.internmc.facebook.com/intern/diff/D18575011)

suo committed:
Re-unify module and interface in ConcreteModuleType

We made `ModuleInfo` a union initially to represent the idea that a
submodule could either be a regular module or a module interface.

This PR represents module interfaces as a ConcreteModuleType with no
info (e.g.  no "sugaredness"), and with the interface type as the
underlying `jitType_`. This has the effect of reducing the special
casing around adding/maintaining module info.

suo commented:
cc @wanchaol

kostmo commented:
## CircleCI build failures summary
As of commit 496e993:
* **5/5** failures introduced in this PR
* **5/5** recognized as flaky
    * Re-run these jobs?



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=496e993f981a1161da33bccd18458f7e00d26f1e):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_linux_xenial_py3_6_gcc5_4_build | Set Up CI Environment After attach_workspace | `E: Failed to fetch` |
| pytorch_linux_xenial_py2_7_9_build | Set Up CI Environment After attach_workspace | `E: Failed to fetch` |
| pytorch_xla_linux_xenial_py3_6_clang7_build | Set Up CI Environment After attach_workspace | `E: Failed to fetch` |
| pytorch_linux_xenial_py3_clang5_android_ndk_r19c_x86_32_build | Set Up CI Environment After attach_workspace | `E: Failed to fetch` |
| pytorch_linux_xenial_cuda9_cudnn7_py3_build | Set Up CI Environment After attach_workspace | `E: Failed to fetch` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

wanchaol reviewed (approved):
looks nice! one minor comment.

suo committed:
Update on "Re-unify module and interface in ConcreteModuleType"


We made `ModuleInfo` a union initially to represent the idea that a
submodule could either be a regular module or a module interface.

This PR represents module interfaces as a ConcreteModuleType with no
info (e.g.  no "sugaredness"), and with the interface type as the
underlying `jitType_`. This has the effect of reducing the special
casing around adding/maintaining module info.

zdevito commented on torch/csrc/jit/script/python_sugared_value.cpp:
```diff
@@ -266,23 +266,23 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
-  const auto& selfType = concreteType_->getJitType();
+  const auto& selfType_ = concreteType_->getJitType();
+  if (selfType_->cast<InterfaceType>()) {
+    return std::make_shared<SimpleValue>(self_)->attr(loc, m, field);
```
`SimpleValue(self)->` since we are deleting the object directly after.

suo replied:
Changing it would cause a lint error internally: https://fburl.com/wiki/jshlq9g3. Which doesn't apply in this situation but I guess could conceivably happen with some more code mutations down the road, so we might as well follow the rule?

zdevito commented on torch/csrc/jit/script/python_sugared_value.cpp:
```diff
@@ -266,23 +266,23 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
-      return std::make_shared<SimpleValue>(self_)->attr(loc, m, field);
+    // ...otherwise, methods, parameters, attributes, and buffers are all
+    // first class so they get returned as SimpleValues
+    return std::make_shared<SimpleValue>(self_)->attr(loc, m, field);
```
same here

suo committed:
Update on "Re-unify module and interface in ConcreteModuleType"


We made `ModuleInfo` a union initially to represent the idea that a
submodule could either be a regular module or a module interface.

This PR represents module interfaces as a ConcreteModuleType with no
info (e.g.  no "sugaredness"), and with the interface type as the
underlying `jitType_`. This has the effect of reducing the special
casing around adding/maintaining module info.

Differential Revision: [D18575011](https://our.internmc.facebook.com/intern/diff/D18575011)

suo committed:
Update on "Re-unify module and interface in ConcreteModuleType"


We made `ModuleInfo` a union initially to represent the idea that a
submodule could either be a regular module or a module interface.

This PR represents module interfaces as a ConcreteModuleType with no
info (e.g.  no "sugaredness"), and with the interface type as the
underlying `jitType_`. This has the effect of reducing the special
casing around adding/maintaining module info.

Differential Revision: [D18575011](https://our.internmc.facebook.com/intern/diff/D18575011)
